cores/esp8266/WString.h

```diff
@@ -343,48 +345,85 @@ class String {
         void invalidate(void);
         unsigned char changeBuffer(unsigned int maxStrLen);
 
-        // copy and move
+        // copy or insert at a specific position
         String &copy(const char *cstr, unsigned int length);
         String &copy(const __FlashStringHelper *pstr, unsigned int length);
+
+        String &insert(size_t position, char);
+        String &insert(size_t position, const char *);
+        String &insert(size_t position, const __FlashStringHelper *);
+        String &insert(size_t position, const char *, size_t length);
+        String &insert(size_t position, const String &);
+
+        // rvalue helper
         void move(String &rhs) noexcept;
 };
 
-class StringSumHelper: public String {
-    public:
-        StringSumHelper(const String &s) :
-                String(s) {
-        }
-        StringSumHelper(const char *p) :
-                String(p) {
-        }
-        StringSumHelper(char c) :
-                String(c) {
-        }
-        StringSumHelper(unsigned char num) :
-                String(num) {
-        }
-        StringSumHelper(int num) :
-                String(num) {
-        }
-        StringSumHelper(unsigned int num) :
-                String(num) {
-        }
-        StringSumHelper(long num) :
-                String(num) {
-        }
-        StringSumHelper(unsigned long num) :
-                String(num) {
-        }
-        StringSumHelper(float num) :
-                String(num) {
-        }
-        StringSumHelper(double num) :
-                String(num) {
-        }
-        StringSumHelper(const __FlashStringHelper *s) :
-                String(s) {
-        }
-};
+// concatenation (note that it's done using non-method operators to handle both possible type refs)
+
+inline String operator +(const String &lhs, const String &rhs) {
+    String res;
+    res.reserve(lhs.length() + rhs.length());
+    res += lhs;
+    res += rhs;
+    return res;
+}
+
+inline String operator +(String &&lhs, const String &rhs) {
+    lhs.concat(rhs);
+    return std::move(lhs);
+}
+
+String operator +(const String &lhs, String&& rhs);
+String operator +(String &&lhs, String &&rhs);
+
+// concat / += already handles the basic types, so just do that
+// TODO: could also do `operator+=` as a templated method
+// TODO: explicitly list concat overloads in the enable_if?
+
+template <typename T>
+inline std::enable_if_t<std::is_pod_v<T> && !std::is_same_v<std::decay<T>, String>, String>
+operator +(const String &lhs, T value) {
+    String res(lhs);
+    res += value;
+    return res;
+}
+
+template <typename T>
+inline std::enable_if_t<std::is_pod_v<T> && !std::is_same_v<std::decay<T>, String>, String>
+operator +(String &&lhs, T value) {
+    lhs += value;
+    return std::move(lhs);
+}
```

Comment:
Explicit option I mention in the comment is something like
```cpp
template <typename T, typename R>
using enable_if_string_concat_type = typename std::enable_if<std::is_same_v<int, T> || std::is_same_v<unsigned int, T> || ... ), R>;
// the rest of the concat overloads
```

There's also [`std::is_detected<CheckType, T>`](https://en.cppreference.com/w/cpp/experimental/is_detected), or template specialization could be explicit and only allow specific `T`s

Or, don't use any templates / SFINAE and write every overload like the `String& operator+=(...)` already does
(removes type_traits include, and no templates to parse for the compiler)